src/dune_pkg/lock_dir.ml

```diff
@@ -46,6 +46,61 @@ module Source = struct
         ]
   ;;
 
+  let fetch_and_hash_archive url =
+    let open Fiber.O in
+    let temp_dir = Temp.create Dir ~prefix:"dune" ~suffix:"archive" in
+    Fiber.finalize ~finally:(fun () ->
+      Temp.destroy Dir temp_dir;
+      Fiber.return ())
+    @@ fun () ->
+    let target = Path.relative temp_dir "archive" in
+    Fetch.fetch ~unpack:false ~checksum:None ~target (OpamUrl.of_string url)
+    >>| function
+    | Ok () -> Some (Dune_digest.file target |> Checksum.of_dune_digest)
```

Comment:
FYI, the internals of `Fetch.fetch` can already give you the newly calculated checksum. It should be enough to just tweak the API to get it somehow.